src/Cryptography/Sodium/Bindings/SHA512.hs

```diff
@@ -0,0 +1,122 @@
+{-# LANGUAGE CApiFFI #-}
+{-# LANGUAGE Trustworthy #-}
+
+-- |
+--
+-- Module: Cryptography.Sodium.Bindings.SHA512
+-- Description: Direct bindings to the SHA-512 hashing function implemented in Libsodium
+-- Copyright: (C) Hécate Moonlight 2022
+-- License: BSD-3-Clause
+-- Maintainer: The Haskell Cryptography Group
+-- Stability: Stable
+-- Portability: GHC only
+module Cryptography.Sodium.Bindings.SHA512
+  ( -- * Introduction
+    -- $introduction
+
+    -- * Single-part message
+    cryptoHashSHA512,
+
+    -- * Multi-part messages
+    CryptoHashSHA512State (..),
+    cryptoHashSHA512Init,
+    cryptoHashSHA512Update,
+    cryptoHashSHA512Final,
+    -- * Constants
+    cryptoHashSHA512Bytes,
+  )
+where
+
+import Foreign (Ptr, Word64, Word8)
+import Foreign.C (CInt (..), CSize (..), CUChar (..), CULLong (..))
+
+-- $introduction
+--
+-- The SHA-512 functions are provided for interoperability with other applications. If you are
+-- looking for a generic hash function and not specifically SHA-2, using
+-- 'Cryptography.Sodium.Bindings.GenericHashing' (BLAKE2b) might be a better choice.
+-- These functions are also not suitable for hashing passwords or deriving keys from passwords.
+-- Use 'Cryptography.Sodium.Bindings.PasswordHashing' instead.
+--
+-- These functions are not keyed and are thus deterministic. In addition, the untruncated versions
+-- are vulnerable to length extension attacks. A message can be hashed in a single pass, but a
+-- streaming API is also available to process a message as a sequence of multiple chunks.
+
+
+-- | Hash the content of the @in@ buffer and put the result in the @out@ buffer.
+-- The @inlen@ parameter is the size of what is in the @in@ buffer.
+--
+-- @since 0.0.1.0
+foreign import capi "sodium.h crypto_hash_sha512"
+  cryptoHashSHA512 ::
+    -- | @out@ parameter. This is a pointer to the hash of your data.
+    Ptr CUChar ->
+    -- | @in@ parameter. This is a pointer to the data you want to hash.
+    Ptr CUChar ->
+    -- | @inlen@ parameter. This is the length of the data you want to hash.
+    CULLong ->
+    -- | Return code is 0 on success, -1 on error.
+    IO CInt
+
+-- | This is the state held and used by the various functions of this
+-- module.
+--
+-- @since 0.0.1.0
+data CryptoHashSHA512State = CryptoHashSHA512State
```

Comment:
If we're going for an 'owning' representation like this, it needs a `Storable` instance.

Furthermore, how do you know this is the structure? It's an opaque type as far as I can tell from the docs.

Reply:
Empty data type is probably what you want here. We also need a constant defining what size this is, or people won't be able to allocate it.